ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/RestoreCommandsCompactor.java

```diff
@@ -50,73 +45,92 @@ private RestoreCommandsCompactor() {
+    final List<QueuedCommand> compacted = new LinkedList<>();
+    while (current != null) {
+      // traverse backwards and add each next node to the start of the list
+      compact(current).ifPresent(cmd -> compacted.add(0, cmd));
+      current = current.prev;
+    }
 
     return compacted;
   }
 
-  private static Set<QueuedCommand> findTerminatedQueriesAndRemoveTerminateCommands(
-      final List<QueuedCommand> commands
-  ) {
-    final Map<QueryId, QueuedCommand> queries = new HashMap<>();
-    final Set<QueuedCommand> terminatedQueries = Collections.newSetFromMap(new IdentityHashMap<>());
-
-    final Iterator<QueuedCommand> it = commands.iterator();
-    while (it.hasNext()) {
-      final QueuedCommand cmd = it.next();
-
-      // Find known queries:
-      final Command command =
-          cmd.getAndDeserializeCommand(InternalTopicSerdes.deserializer(Command.class));
-      if (command.getPlan().isPresent()
-          && command.getPlan().get().getQueryPlan().isPresent()
-      ) {
-        final QueryId queryId =
-            command.getPlan().get().getQueryPlan().get().getQueryId();
-        queries.putIfAbsent(queryId, cmd);
+  private static final class CompactedNode {
+
+    final CompactedNode prev;
+    final QueuedCommand queued;
+    final Command command;
+
+    boolean shouldSkip = false;
+
+    public static CompactedNode of(
+        final CompactedNode prev,
+        final QueuedCommand queued,
+        final Map<QueryId, CompactedNode> latestNodeWithId
+    ) {
+      final Command command = queued.getAndDeserializeCommand(
+          InternalTopicSerdes.deserializer(Command.class)
+      );
+
+      final Optional<KsqlPlan> plan = command.getPlan();
+      if (queued.getAndDeserializeCommandId().getType() == Type.TERMINATE) {
+        final QueryId queryId = new QueryId(queued.getAndDeserializeCommandId().getEntity());
+        markShouldSkip(queryId, latestNodeWithId);
+
+        // terminate commands don't get added to the list of commands to execute
+        // because we "execute" them in this class by removing query plans from
+        // terminated queries
+        return prev;
+      } else if (!plan.isPresent() || !plan.get().getQueryPlan().isPresent()) {
+        // DDL
+        return new CompactedNode(prev, queued, command);
       }
 
-      // Find TERMINATE's that match known queries:
-      if (cmd.getAndDeserializeCommandId().getType() == Type.TERMINATE) {
-        final QueryId queryId = new QueryId(cmd.getAndDeserializeCommandId().getEntity());
-        final QueuedCommand terminated = queries.remove(queryId);
-        if (terminated != null) {
-          terminatedQueries.add(terminated);
-          it.remove();
-        }
-      }
-    }
+      final QueryId queryId = plan.get().getQueryPlan().get().getQueryId();
+      markShouldSkip(queryId, latestNodeWithId);
+      final CompactedNode node = new CompactedNode(prev, queued, command);
```

Comment:
So, you create a compacted node even for the previous queries that have been replaced?

Reply:
yes, we create the node when we hit the command because we don't know at the time whether or not it will be compacted. later, when we see another node with the same queryId, we mark the older one as compacted (`shouldSkip`)

Reply:
Ah, but every time you replace it when you find a new one. OK.